getline.ts/src/blockchain.ts

```diff
@@ -0,0 +1,237 @@
+import * as Web3 from 'web3';
+import {BigNumber} from 'bignumber.js';
+import * as moment from 'moment';
+
+import {Address} from './common';
+import {MetabackendClient, pb} from './metabackend';
+
+/**
+ * A connector used to access blockchains.
+ */
+export interface Blockchain {
+    /**
+     * Convert a given block number in the future to a timestamp when it lands
+     * on the blockchain.
+     * @param current Current block number.
+     * @param target Block number in the future.
+     * @returns Timestamp of target block number.
+     */
+    blockToTime(current: BigNumber, target: BigNumber): moment.Moment;
+    /**
+     * Returns current top block number on the blockchain.
+     * @returns Block number.
+     */
+    currentBlock(): Promise<BigNumber>;
+    /**
+     * Retrieves a handle to a contract at an address on the blockchain.
+     * @param name Name of contract on the metabackend.
+     * @param address Address of contract on the blockchain.
+     * @returns Contract at this address.
+     */
+    existing(name: string, address: Address): Promise<Contract>;
+    /**
+     * Deploys a new contract on the blockchain.
+     * @param name Name of contract on the metabackend.
+     * @param params Constructor parameters of the contract.
+     * @returns Newly deployed contract.
+     */
+    deploy(name: string, ...params: Array<any>): Promise<Contract>;
+    /**
+     * Returns coinbase of the blockchain client.
+     * @returns Coinbase address.
+     */
+    coinbase(): Promise<Address>;
+}
+
+/**
+ * A contract deployed on a blockchain.
+ */
+export class Contract {
+    /**
+     * Address of the contract on the blockchain.
+     */
+    public address: Address
+    /**
+     * Blockchain on which the contract is located.
+     */
+    private blockchain: Blockchain
+    private instance: Web3.ContractInstance
+
+    constructor(blockchain: Blockchain, instance: Web3.ContractInstance) {
+        this.blockchain = blockchain;
+        this.instance = instance;
+        this.address = new Address(this.blockchain, instance.address);
+    }
+
+    /**
+     * Call a contract method.
+     * @param methodName Name of method to call.
+     * @param params Arguments to pass to method.
+     * @returns Result of call.
+     */
+    public async call(methodName: string, ...params: Array<any>): Promise<any> {
```

Comment:
Javascript is reflective language, why not wrap all the abi calls and enjoy typing of funciton calls instead of a string call?
See [Truffle-Contract wrapper](https://github.com/trufflesuite/truffle-contract/blob/develop/contract.js#L273)

Reply:
I prefer less magic to more magic, to be honest. Having an explicit .call function lets the programmer know that we're passing over the boundary to a smart contract, with all the problems that this entails (lack of type safety, runtime errors, timeouts, etc).